src/utils.rs

```diff
@@ -11,3 +11,20 @@ pub(crate) const fn as_ptr<T>(t: &T) -> *const T {
 pub(crate) fn as_mut_ptr<T>(t: &mut T) -> *mut T {
     t
 }
+
+/// Convert a `*const c_void` to a `*const T`, checking that it is not null,
+/// misaligned, or pointing to a region of memory that wraps around the address
+/// space.
+#[allow(dead_code)]
+pub(crate) fn check_raw_pointer<T>(value: *const core::ffi::c_void) -> Option<*const T> {
```

Comment:
For my own knowledge, I'm curious if there is any reason not to use `NonNull<T>` here so we could pack the `Option`?

Reply:
Only that it hadn't occurred to me :-). But looking into it, `NonNull` is `*mut` rather than `*const`, which is a little inconvenient in src/backend/linux_raw/vdso.rs where it's used a bunch. I'll need to think about this more.

Reply:
I've now added a patch to do this. It is a little more verbose, but I think it's ok.